src/compiler/moduleNameResolver.ts

```diff
@@ -2400,6 +2404,37 @@ function getVersionPathsOfPackageJsonInfo(packageJsonInfo: PackageJsonInfo, stat
     return packageJsonInfo.contents.versionPaths || undefined;
 }
 
+function getPeerDependenciesOfPackageJsonInfo(packageJsonInfo: PackageJsonInfo, state: ModuleResolutionState): string | undefined {
+    if (packageJsonInfo.contents.peerDependencies === undefined) {
+        packageJsonInfo.contents.peerDependencies = readPackageJsonPeerDependencies(packageJsonInfo, state) || false;
+    }
+    return packageJsonInfo.contents.peerDependencies || undefined;
+}
+
+function readPackageJsonPeerDependencies(packageJsonInfo: PackageJsonInfo, state: ModuleResolutionState): string | undefined {
+    const peerDependencies = readPackageJsonField(packageJsonInfo.contents.packageJsonContent, "peerDependencies", "object", state);
+    if (peerDependencies === undefined) return undefined;
+    if (state.traceEnabled) trace(state.host, Diagnostics.package_json_has_a_peerDependencies_field);
+    const packageDirectory = realPath(packageJsonInfo.packageDirectory, state.host, state.traceEnabled);
+    const nodeModules = packageDirectory.substring(0, packageDirectory.lastIndexOf("node_modules") + "node_modules".length) + directorySeparator;
+    let result = "";
+    for (const key in peerDependencies) {
+        if (hasProperty(peerDependencies, key)) {
+            const peerPackageJson = getPackageJsonInfo(nodeModules + key, /*onlyRecordFailures*/ false, state);
+            if (peerPackageJson) {
+                const version = (peerPackageJson.contents.packageJsonContent as PackageJson).version;
+                result += `+${key}@${version}`;

```

Comment:
Q: I could theoretically name a package on disk `typescript@5.4.1+` and munge this cache key up pretty badly (all these separators are valid filepath characters after all). Shouldn't this use, eg, `?` and `|` instead of `+` and `@`, since, at least on windows, those aren't valid path characters? Believe it or not, a `require("+typescript@5.4.1")` will work and resolve on disk, even if `npm` won't let you publish a package with those characters!

Reply:
eg. if your file has import whose peer dependency is the source File for the package then short cut to directly use package name?

Reply:
```ts 
import { a } from "foo";
export functopn bar() { 
return a.bar(); // This is from package Bar@BarVersion
}
```
then while going through all imports in the file: if you find say "foo" import whose peer dependency is "Bar@Bar.version" use that package name ?

Reply:
Maybe that could be leveraged somehow, but getting just the package name is usually fairly easy compared to figuring out what subpath can be used to get to the export we need.

Reply:
@weswigham @andrewbranch  do we still have concern about the string for peer dependency calculations.  
Do we want this as array of package name and versions for future use or just some string is ok. Should i make this internal for now so we have option to change this later? Would like to get this in sooner than later

Reply:
I think it should probably be internal if it’s a formatted string that only gets used as a lookup id. But if you want to go ahead and change it to be an object with name/version, that works too.